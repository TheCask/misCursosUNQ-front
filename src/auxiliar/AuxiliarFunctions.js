
export function setInnerPropValue(baseObj, subPropString, value) {
    const subProps = subPropString.split(".");
    const lastPropName = subProps.pop(); // elimina del array y retorna el ultimo 
    let propRef = baseObj
    subProps.forEach(subprop => {
      propRef = propRef[subprop];
    });
    propRef[lastPropName] = value;
}