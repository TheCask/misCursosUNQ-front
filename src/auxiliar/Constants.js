export const userCsvToJsonMap = {
    dni: 'personalData.dni',
    firstName: 'personalData.firstName',
    lastName: 'personalData.lastName',
    email: 'personalData.email',
    cellPhone: 'personalData.cellPhone',
    cuitNumber: 'jobDetail.cuitNumber',
    category: 'jobDetail.category',
    grade: 'jobDetail.grade',
    dedication: 'jobDetail.dedication',
    contractRelation: 'jobDetail.contractRelation',
    aditionalHours: 'jobDetail.aditionalHours',
    cvURL: 'jobDetail.cvURL',
    lastUpdate: 'jobDetail.lastUpdate',
    gradeTitles: 'jobDetail.gradeTitles',
    posGradeTitles: 'jobDetail.posGradeTitles'
};

export const studentsCsvToJsonMap = {
    fileNumber: 'fileNumber',
    dni: 'personalData.dni',
    firstName: 'personalData.firstName', 
    lastName: 'personalData.lastName', 
    email: 'personalData.email',
    cellPhone: 'personalData.cellPhone'
};

export const CategoryOptions = ['', 'Auxiliar', 'Intructor/a', 'Adjunto/a', 'Asociado/a', 'Titular', 'Emérito/a', 'Consulto/a']
export const GradeOptions = ['', 'A', 'B']
export const ContractOptions = ['', 'Contratado/a', 'Interino/a', 'Ordinario/a']
export const DedicationOptions = ['', 'Parcial', 'Semi-Exclusiva', 'Exclusiva']

export const SeasonOptions = ['1C', '2C', '3C', '1T', '2T', '3T', '4T', '1S', '2S']
export const ShiftOptions = ['Mañana', 'Tarde', 'Noche']

export const emptyUser = {
    isActive: true,
    personalData: {
        dni: '',
        firstName: '',
        lastName: '',
        email: '',
        cellPhone: '' 
    },
    jobDetail: {
        cuitNumber: '',
        category: '',
        grade: '',
        dedication: '',
        contractRelation: '',
        aditionalHours: 0,
        cvURL: '',
        lastUpdate: '',
        gradeTitles: '',
        posGradeTitles: ''
    },
    coordinatedSubjects: [],
    taughtCourses: []
};

export const emptyCourse = {
    courseCode: '',
    courseShift: '',
    courseIsOpen: '',
    courseYear: 0,
    courseSeason: '',
    courseLocation: '',
    subject: {
        code: ''
    },
    students: [],
    teachers: []
};

export const emptyNewCourse = {
    courseCode: '',
    courseShift: '',
    courseIsOpen: true,
    courseYear: 2020,
    courseSeason: '',
    courseLocation: '',
    subject: {
        code: '',
        name: ''
    },
    students: [],
    lessons: [],
    teachers: [],
    evaluations: []
};

export const profileUser = {
    birthDate: "",
    data: {}, // not used
    email: "",
    firstName: "",
    fullName: "", // Autoconstruct lastName, firstName
    imageUrl: "",
    lastName: "",
    middleName: "", // Not Used
    mobilePhone: "",
    passwordChangeRequired: false,
    preferredLanguages: [], // this property always concats to the list (PATCH method bug on FusionAuth)
    timezone: "Etc/GMT-3", // not change
    twoFactorDelivery: "None", // not used
    twoFactorEnabled: false, // not used
    usernameStatus: "", // not used
    username: ""
};

export const emptyStudent = {
    fileNumber: 0,
    personalData: {
      // personalDataId: 0,
      dni: 0,
      firstName: '',
      lastName: '',
      email: '',
      cellPhone: ''
    },
    takenCourses: [],
    attendedLessons: []
};

export const emptySubject = {
    code: '',
    name: '',
    acronym: '',
    programURL: ''
};

export const parserConfig = {
    delimiter: "",  // auto-detect
    newline: "",  // auto-detect
    quoteChar: '"',
    escapeChar: '"',
    header: true,
    transformHeader: undefined,
    dynamicTyping: true,
    preview: 0,
    encoding: "",
    worker: true,
    comments: false,
    step: undefined,
    complete: undefined,
    error: undefined,
    download: false,
    downloadRequestHeaders: undefined,
    skipEmptyLines: false,
    chunk: undefined,
    fastMode: undefined,
    beforeFirstChunk: undefined,
    withCredentials: undefined,
    transform: undefined,
    delimitersToGuess: [',', '	', '|', ';']
};

export function newEvalInstance() {
    return {
        evaluationId: '',
        instanceName: 'TP ',
        califications: []
    }
}

export const initialStudent = () => { return { personalData: {} } }

export const initialUser = () => { return { isActive: true, personalData: {}, jobDetail: {} } }