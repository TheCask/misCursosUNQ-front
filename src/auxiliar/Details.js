import React, { Component } from 'react';
import UserEdit from '../user/UserEdit'
import StudentEdit from '../student/StudentEdit'
import SubjectEdit from '../subject/SubjectEdit'
import CourseEdit from '../course/CourseEdit'
import { userContext } from '../login/UserContext';
import AccessError from '../errorHandling/AccessError';

export class UserDetail extends Component {
  render() {
    this.actualRol = this.context.actualRol;
    return (this.actualRol === 'Guest' ?
      <AccessError errorCode="Guests are not allowed" 
          errorDetail="Make sure you are signed in with valid role before try to access this page"/>
      : 
      <UserEdit onlyDetail={true}/>
    )
  }
}

export class StudentDetail extends Component {
  render() {
    this.actualRol = this.context.actualRol;
    return (this.actualRol === 'Guest' ?
      <AccessError errorCode="Guests are not allowed" 
          errorDetail="Make sure you are signed in with valid role before try to access this page"/>
      : 
      <StudentEdit onlyDetail={true}/>
    )
  }
}

export class SubjectDetail extends Component {
  render() {
    this.actualRol = this.context.actualRol;
    return (this.actualRol === 'Guest' ?
      <AccessError errorCode="Guests are not allowed" 
          errorDetail="Make sure you are signed in with valid role before try to access this page"/>
      : 
      <SubjectEdit onlyDetail={true}/>
    )
  }
}

export class CourseDetail extends Component {
  render() {
    this.actualRol = this.context.actualRol;
    return (this.actualRol === 'Guest' ?
      <AccessError errorCode="Guests are not allowed" 
          errorDetail="Make sure you are signed in with valid role before try to access this page"/>
      : 
      <CourseEdit onlyDetail={true}/>
    )
  }
}

UserDetail.contextType = userContext;
StudentDetail.contextType = userContext;
SubjectDetail.contextType = userContext;
CourseDetail.contextType = userContext;