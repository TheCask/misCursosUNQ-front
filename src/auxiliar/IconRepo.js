import React from 'react';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

export default function getIcon(operationType, entityTypeCapName) {
    
    switch(operationType) {

        case 'Add' : 

            switch (entityTypeCapName) {
            
                case 'User': return ( <span>
                    <FontAwesomeIcon icon="chalkboard-teacher" size="1x"/>
                    <FontAwesomeIcon icon="plus-circle" size="1x" transform="right-5 up-5"/>
                </span> )
                case 'Teacher': return ( <span>
                    <FontAwesomeIcon icon="chalkboard-teacher" size="1x"/>
                    <FontAwesomeIcon icon="plus-circle" size="1x" transform="right-5 up-5"/>
                </span> )
                case 'Coordinator': return ( <span>
                    <FontAwesomeIcon icon="chalkboard-teacher" size="1x"/>
                    <FontAwesomeIcon icon="plus-circle" size="1x" transform="right-5 up-5"/>
                </span> )
                case 'Student': return ( <span>
                    <FontAwesomeIcon icon="user-graduate" size="1x"/>
                    <FontAwesomeIcon icon="plus-circle" size="1x" transform="right-5 up-5"/>
                </span> )
                case 'Subject': return ( <span>
                    <FontAwesomeIcon icon="university" size="1x"/>
                    <FontAwesomeIcon icon="plus-circle" size="1x" transform="right-5 up-5"/>
                </span> )
                case 'Course': return ( <span>
                    <FontAwesomeIcon icon="chalkboard" size="1x"/>
                    <FontAwesomeIcon icon="plus-circle" size="1x" transform="right-5 up-5"/>
                </span> )
                default: return (<FontAwesomeIcon icon='plus-circle' size="2x" style={{ width: "1em"}}/>)
            }

        case 'Edit' :

            switch (entityTypeCapName) {

                default : return <FontAwesomeIcon icon='edit' size="2x" style={{ width: "1em"}}/>
            }
            
        case 'Delete' :

            switch (entityTypeCapName) {

                default : return <FontAwesomeIcon icon='trash-alt' size="2x"/>
            }

        default : return (<FontAwesomeIcon icon='plus-circle' size="2x" style={{ width: "1em"}}/>)

    }
}

export function getCourseIcon(subjectCode) {
    switch(subjectCode.split("-")[0]) {
    case "80000":
        return (<> <FontAwesomeIcon icon='book' size="1x" color="darkred" transform="left-10 up-10"/>
                <FontAwesomeIcon icon='book-reader' size="1x" color="darkred" transform="right-10 up-10"/>
                <FontAwesomeIcon icon='pencil-alt' size="1x" color="darkred" transform="left-10 down-10"/>
                <FontAwesomeIcon icon='graduation-cap' size="1x" color="darkred" transform="right-10 down-10"/> </>);
    case "80005":
        return (<> <FontAwesomeIcon icon='bug' size="1x" color="black" transform="left-10 up-10"/>
                <FontAwesomeIcon icon='microchip' size="1x" color="black" transform="right-10 up-10"/>
                <FontAwesomeIcon icon='laptop-code' size="1x" color="black" transform="left-10 down-10"/>
                <FontAwesomeIcon icon='project-diagram' size="1x" color="black" transform="right-10 down-10"/> </>);
    case "80003":
        return (<> <FontAwesomeIcon icon='brain' size="1x" color="darkblue" transform="left-10 up-10"/>
                <FontAwesomeIcon icon='shapes' size="1x" color="darkblue" transform="right-10 up-10"/>
                <FontAwesomeIcon icon='infinity' size="1x" color="darkblue" transform="left-10 down-10"/>
                <FontAwesomeIcon icon='calculator' size="1x" color="darkblue" transform="right-10 down-10"/> </>);
    case "80004":
        return (<> <FontAwesomeIcon icon='thermometer-half' size="1x" color="darkgreen" transform="left-10 up-10"/>
                <FontAwesomeIcon icon='atom' size="1x" color="darkgreen" transform="right-10 up-10"/>
                <FontAwesomeIcon icon='flask' size="1x" color="darkgreen" transform="left-10 down-10"/>
                <FontAwesomeIcon icon='magnet' size="1x" color="darkgreen" transform="right-10 down-10"/> </>);
    default: return <FontAwesomeIcon icon={['fas', 'chalkboard']} size="2x" color="gray"/>
    }
}