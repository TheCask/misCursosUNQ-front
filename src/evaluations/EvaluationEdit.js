import React, { useState } from 'react';
import {  Input, InputGroup, InputGroupAddon,ButtonGroup, Badge} from 'reactstrap';
import Col from 'reactstrap/lib/Col';
import Row from 'reactstrap/lib/Row';
import { Button, Table } from 'reactstrap';
import Log from '../auxiliar/Log';
import CRUDEditButton from '../CRUDButtonBar/CRUDEditButton';
import CRUDDeleteButton from '../CRUDButtonBar/CRUDDeleteButton';
import * as CourseAPI from '../services/CourseAPI';


export default function EvaluationEdit(props){

    const [filterValue, setFilterValue] = useState('');

    function newEvalInstance() {
        return {
            evaluationId: '',
            instanceName: 'TP ',
            califications: []
        }
    }

    Log.info("EvaEdit");

    /* useEffect(() => {
        setEvaluations(props.evaluations);
    }, [props.evaluations]); */

    function setRowColor (itemId) {
        if (props.currEvalInstance.evaluationId === itemId) {
        return {backgroundColor:'#F0FFF0'}
        }
    }

    function onClickFunc(evalId){
        return () => {
            props.onSelectFunc(evalId);
        }
    }

    function handleEvalSubmit(event, evalInstance) {
        event.preventDefault();
        CourseAPI.postCourseEvaluationAsync(
            props.courseId, 
            evalInstance, 
            (res) => {
                Log.info(res, "POST SUCCEDED");
                props.reloadCourse();     /// NEEDED??????
            }, 
            props.showError("save evaluation"));
    }

    function handleEvalDelete(event, evalInstance) {
        event.preventDefault();
        CourseAPI.deleteCourseEvaluationAsync(
            evalInstance.evaluationId, 
            (res) => {
                Log.info(res, "DELETE SUCCEDED");
                props.reloadCourse();     /// NEEDED??????
            }, 
            props.showError("delete evaluation"));
    }

    function getInnerPropValue(baseObj, subPropString){
        const subProps = subPropString.split(".");
        const lastPropName = subProps.pop(); // elimina del array y retorna el ultimo 
        let propRef = baseObj;
        subProps.forEach(subprop => {
          propRef = propRef[subprop];
        });
        return propRef[lastPropName];
    }

    function isEvalSelected() { 
        return props.currEvalInstance && props.currEvalInstance.evaluationId !== '';
    }

    const tableRender = (evalList) => {

        const filteredFields = ["instanceName"];
        let filteredList = evalList;
        if (filterValue !== '')
            filteredList = evalList.filter( ev => 
                filteredFields.some( k => {
                    Log.info(getInnerPropValue(ev, k), "getInnerPropValue(ev, k)");
                    return getInnerPropValue(ev, k) ? getInnerPropValue(ev, k).toString().toLowerCase().includes(filterValue.toLowerCase()) : false
                
                })             
        )

        function isValidNote(note){
            return note && !isNaN(note);
        }

        function countValidNotes(califications){
            return califications.reduce((acc, currCal)=> isValidNote(currCal.note) ? (acc + 1) : acc, 0)
        }

        function sumValidNotes(califications){
            return califications.reduce((acc, currCal)=> isValidNote(currCal.note) ? (acc + currCal.note) : acc, 0)
        }

        function avgValidNotes(califications){
            return (sumValidNotes(califications)/countValidNotes(califications)).toFixed(1)
        }

        const tdStyle = {whiteSpace: 'nowrap', textAlign: 'center'};

        return (
            <Table hover className="mt-4" >
                <thead>
                    <tr>
                        <th width="100%">Instance Name</th>
                        <th {...tdStyle} width="100%">Average</th>
                        <th {...tdStyle} width="100%">Califs.</th>
                    </tr>
                </thead>
                <tbody>
                    {filteredList.map( (ev, index) => 
                        <tr style={setRowColor(ev.evaluationId)}
                            key={index} 
                            id={ev.evaluationId} 
                            onClick={onClickFunc(ev.evaluationId)} 
                        >
                            <td style={{whiteSpace: 'nowrap'}}>{ev.instanceName || ''}</td>
                            <td style={tdStyle}>
                                {avgValidNotes(ev.califications)}
                            </td>
                            <td style={tdStyle}>
                                <Badge color="primary" pill>
                                    {`${countValidNotes(ev.califications)}/${ev.califications.length}`}
                                </Badge>
                            </td>
                        </tr>
                    )}
                </tbody>
            </Table>)
    };

    return (
        <>
        <Row>
            <Col md={9}>
                <InputGroup>
                    <Input 
                        placeholder={"Search evaluation by name"} 
                        value={filterValue} 
                        onChange={(event) => setFilterValue(event.target.value)} 
                        disabled = {props.evaluations.length === 0}
                        />
                    <InputGroupAddon addonType="append">
                        <Button color="secondary" onClick={() => setFilterValue('')} >Clear</Button>
                    </InputGroupAddon>
                </InputGroup>
            </Col>
            <Col>
                <ButtonGroup style={{ padding: "0em", float: "right"}}>
                    <CRUDEditButton 
                        operationType='Add'
                        behavior = {{ onProceed: handleEvalSubmit }}  // requires either {onClick: <myFunc>} or {to: <'/my/link'>' tag={Link}} 
                        onClick = {() => {}}
                        entityTypeCapName = {'Evaluation'}
                        isDisabled = {false}
                        getEntity = {newEvalInstance()}
                        />
                    <CRUDEditButton 
                        operationType='Edit'
                        behavior = {{ onProceed: handleEvalSubmit }}  // requires either {onClick: <myFunc>} or {to: <'/my/link'>' tag={Link}} 
                        onClick = {() => {}}
                        entityTypeCapName = {'Evaluation'}
                        isDisabled = {!isEvalSelected()}
                        getEntity = {props.currEvalInstance}
                        />
                    <CRUDDeleteButton
                        operationType='Delete'
                        behavior = {{ onProceed: handleEvalDelete }}  // requires either {onClick: <myFunc>} or {to: <'/my/link'>' tag={Link}} 
                        onClick = {() => {}}
                        entityTypeCapName = {'Evaluation'}
                        consequences={['Every student calification corresponding to this evaluation instance will be permanently deleted.', "This evaluation will be deleted."]}
                        isDisabled = {!isEvalSelected()}
                        getEntity = {props.currEvalInstance}
                        />
                </ButtonGroup>
            </Col>
        </Row>






        
        {/* <EvaluationPicker 
            evaluations = {currCourse.evaluations}
            onSelectFunc = {setCurrEvalInstanceById}
            currEvalInstance = {currEvalInstance}
            /> */}

        {tableRender(props.evaluations)}
        </>
    )
}