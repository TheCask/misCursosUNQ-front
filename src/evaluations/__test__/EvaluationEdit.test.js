import React from 'react';
import {render, fireEvent, cleanup } from '@testing-library/react';
import EvaluationEdit from '../EvaluationEdit';
import * as Constants from '../../auxiliar/Constants';


afterEach(cleanup);

it("renders correctly when no evaluations are provided", () => {
    const {queryByTestId} = render(<EvaluationEdit 
        courseId={1} //props.courseId}
        evaluations = {[]} //currCourse.evaluations}
        studentQty = {2} //currCourse.students.length}
        currEvalInstance = {Constants.newEvalInstance()} //currEvalInstance}
        onSelectFunc = {null} //setCurrEvalInstanceById}
        reloadCourse={null} //reloadCourse}
        showError={null} //props.showError}
        
        />
    )
    expect(queryByTestId("evalTable")).toBeTruthy()
    expect(queryByTestId("evalTableRow_0")).toBeFalsy()
})
    
it("renders correctly when there's one evaluation and 2 students", () => {
    const {queryByTestId} = render(<EvaluationEdit 
        courseId={1} //props.courseId}
        evaluations = {[
            {
                evaluationId: '1',
                instanceName: 'TP 1',
                califications: []
            }
        ]} //currCourse.evaluations}
        studentQty = {2} //currCourse.students.length}
        currEvalInstance = {Constants.newEvalInstance()} //currEvalInstance}
        onSelectFunc = {null} //setCurrEvalInstanceById}
        reloadCourse={null} //reloadCourse}
        showError={null} //props.showError}
        
        />
    )
    expect(queryByTestId("evalTable")).toBeTruthy()
    expect(queryByTestId("evalTableRow_0")).toBeTruthy()
    expect(queryByTestId("evalTableRow_1")).toBeFalsy()

})


it("renders correctly when there are two evaluations and 2 students with no califications", () => {
    const {queryByTestId, asFragment} = render(<EvaluationEdit 
        courseId={1} //props.courseId}
        evaluations = {[
            {
                evaluationId: '1',
                instanceName: 'TP 1',
                califications: []
            },
            {
                evaluationId: '2',
                instanceName: 'TP 2',
                califications: []
            }

        ]} //currCourse.evaluations}
        studentQty = {2} //currCourse.students.length}
        currEvalInstance = {Constants.newEvalInstance()} //currEvalInstance}
        onSelectFunc = {null} //setCurrEvalInstanceById}
        reloadCourse={null} //reloadCourse}
        showError={null} //props.showError}
        
        />
    )
    expect(asFragment()).toMatchSnapshot()
    expect(queryByTestId("evalTable")).toBeTruthy()
    expect(queryByTestId("evalTableRow_0")).toBeTruthy()
    expect(queryByTestId("evalTableRow_1")).toBeTruthy()
    expect(queryByTestId("evalTableRow_2")).toBeFalsy()
    expect(queryByTestId("evalTableRowAvg_0").firstChild.nodeValue).toBe("-")
    
})
        