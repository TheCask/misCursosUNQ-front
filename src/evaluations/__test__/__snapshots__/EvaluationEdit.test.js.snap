// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there are two evaluations and 2 students with no califications 1`] = `
<DocumentFragment>
  <div
    class="row"
  >
    <div
      class="col-md-9"
    >
      <div
        class="input-group"
      >
        <input
          class="form-control"
          placeholder="Search evaluation by name"
          type="text"
          value=""
        />
        <div
          class="input-group-append"
        >
          <button
            class="btn btn-secondary"
            type="button"
          >
            Clear
          </button>
        </div>
      </div>
    </div>
    <div
      class="col"
    >
      <div
        class="btn-group"
        role="group"
        style="padding: 0em; float: right;"
      >
        <div>
          <button
            class="btn btn-success btn-sm"
            id="Add_Evaluation"
            type="button"
          />
        </div>
        <div>
          <button
            class="btn btn-primary btn-sm disabled"
            disabled=""
            id="Edit_Evaluation"
            type="button"
          />
        </div>
        <div>
          <button
            class="btn btn-danger btn-sm disabled"
            disabled=""
            id="Delete_Evaluation"
            type="button"
          />
        </div>
      </div>
    </div>
  </div>
  <table
    class="mt-4 table table-hover"
    data-testid="evalTable"
  >
    <thead>
      <tr>
        <th
          width="100%"
        >
          Instance Name
        </th>
        <th
          width="100%"
        >
          Average
        </th>
        <th
          width="100%"
        >
          Califs.
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        data-testid="evalTableRow_0"
        id="1"
      >
        <td
          style="white-space: nowrap;"
        >
          TP 1
        </td>
        <td
          data-testid="evalTableRowAvg_0"
          style="white-space: nowrap; text-align: center;"
        >
          -
        </td>
        <td
          style="white-space: nowrap; text-align: center;"
        >
          <span
            class="badge badge-primary badge-pill"
          >
            0/2
          </span>
        </td>
      </tr>
      <tr
        data-testid="evalTableRow_1"
        id="2"
      >
        <td
          style="white-space: nowrap;"
        >
          TP 2
        </td>
        <td
          data-testid="evalTableRowAvg_1"
          style="white-space: nowrap; text-align: center;"
        >
          -
        </td>
        <td
          style="white-space: nowrap; text-align: center;"
        >
          <span
            class="badge badge-primary badge-pill"
          >
            0/2
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
